### PR TITLE
Fix: Company Fields variables are not rendered when used in Dynamic Web Content

### LINKED ISSUE #13084: Company Fields variables are not rendered when used in Dynamic Web Content
### Mautic Version

4.4.x series

### PHP version

7.4

### What browsers are you seeing the problem on?

Firefox, Chrome

### What happened?

According to the documentation, Mautic should be able to render variables for company fields just as for lead fields:
https://docs.mautic.org/en/setup/variables#company-contact-fields

However, when using variables for Company Fields in a Dynamic Web Content, the content is not rendered and thus not displayed. When using a contact-field instead it works as expected. 

Also, there has been an unresolved entry in the forum for a similar issue: 
https://forum.mautic.org/t/company-token-custom-fields-not-displaying-in-dynamic-content-block/29237

### How can we reproduce this issue?

Step 1: Create a Company and fill out fields like phone, website etc.
Step 2: Create a contact and add the contact to the company
Step 3: Create a non-campaign based DWC with a variable for a company field e.g. `{contactfield=companyphone}` in the content. Add a filter that is fulfilled by the contact from Step 2 eg. `Email = not empty`
Step 4: Create a landingpage and add the DWC into the LP
Step 5: Open Landingpage for the contact from Step 2 and see that no content is rendered.

### Relevant log output

```shell
[2023-12-18 12:23:53] mautic.NOTICE: PHP Notice - Undefined index: log - in file /var/www/mautic/htdocs/app/bundles/DynamicContentBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 11 {"mauticTemplateVars":{"app":"[object] (Symfony\\Bundle\\FrameworkBundle\\Templating\\GlobalVariables: {})","event":{"event":"dynamic.content.sent","eventId":"dynamic.content.sent9","eventLabel":{"label":"DWC1","href":"/s/dwc/view/1"},"eventType":"Dynamic Content Sent","timestamp":"2023-12-18 12:23:27","extra":{"stat":{"dynamic_content_id":"1","id":"9","dateSent":"2023-12-18 12:23:27","name":"DWC1","sentDetails":[]},"type":"sent"},"contentTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","icon":"fa-envelope","contactId":"292"},"lead":"[object] (Mautic\\LeadBundle\\Entity\\Lead: Mautic\\LeadBundle\\Entity\\Lead with ID #292)","mauticContent":"lead","mauticTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","view":"[object] (Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine: {})"},"app":"[object] (Symfony\\Bundle\\FrameworkBundle\\Templating\\GlobalVariables: {})","event":{"event":"dynamic.content.sent","eventId":"dynamic.content.sent9","eventLabel":{"label":"DWC1","href":"/s/dwc/view/1"},"eventType":"Dynamic Content Sent","timestamp":"2023-12-18 12:23:27","extra":{"stat":{"dynamic_content_id":"1","id":"9","dateSent":"2023-12-18 12:23:27","name":"DWC1","sentDetails":[]},"type":"sent"},"contentTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","icon":"fa-envelope","contactId":"292"},"lead":"[object] (Mautic\\LeadBundle\\Entity\\Lead: Mautic\\LeadBundle\\Entity\\Lead with ID #292)","mauticContent":"lead","mauticTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","view":"[object] (Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine: {})"} 

[2023-12-18 12:23:53] mautic.NOTICE: PHP Notice - Trying to access array offset on value of type null - in file /var/www/mautic/htdocs/app/bundles/DynamicContentBundle/Views/SubscribedEvents/Timeline/index.html.php - at line 11 {"mauticTemplateVars":{"app":"[object] (Symfony\\Bundle\\FrameworkBundle\\Templating\\GlobalVariables: {})","event":{"event":"dynamic.content.sent","eventId":"dynamic.content.sent9","eventLabel":{"label":"DWC1","href":"/s/dwc/view/1"},"eventType":"Dynamic Content Sent","timestamp":"2023-12-18 12:23:27","extra":{"stat":{"dynamic_content_id":"1","id":"9","dateSent":"2023-12-18 12:23:27","name":"DWC1","sentDetails":[]},"type":"sent"},"contentTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","icon":"fa-envelope","contactId":"292"},"lead":"[object] (Mautic\\LeadBundle\\Entity\\Lead: Mautic\\LeadBundle\\Entity\\Lead with ID #292)","mauticContent":"lead","mauticTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","view":"[object] (Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine: {})"},"app":"[object] (Symfony\\Bundle\\FrameworkBundle\\Templating\\GlobalVariables: {})","event":{"event":"dynamic.content.sent","eventId":"dynamic.content.sent9","eventLabel":{"label":"DWC1","href":"/s/dwc/view/1"},"eventType":"Dynamic Content Sent","timestamp":"2023-12-18 12:23:27","extra":{"stat":{"dynamic_content_id":"1","id":"9","dateSent":"2023-12-18 12:23:27","name":"DWC1","sentDetails":[]},"type":"sent"},"contentTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","icon":"fa-envelope","contactId":"292"},"lead":"[object] (Mautic\\LeadBundle\\Entity\\Lead: Mautic\\LeadBundle\\Entity\\Lead with ID #292)","mauticContent":"lead","mauticTemplate":"MauticDynamicContentBundle:SubscribedEvents\\Timeline:index.html.php","view":"[object] (Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine: {})"}
```

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by shinde-rahul**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 4.4
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #13084

#### Description:

M5 version: https://github.com/mautic/mautic/pull/13198

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
How can we reproduce this issue?
2. Create a Company and fill out fields like phone, website etc.
3. Create a contact and add the contact to the company
4. Create a non-campaign-based DWC with a variable for a company field, e.g. {contactfield=companyphone} in the content. Add a filter that is fulfilled by the contact form Step 2 eg. Email = not empty
5. Create a landing page and add the DWC to the LP
6. Open the Landing page for the contact from Step 2 and see that no content is rendered.

### CODE REVIEW

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php [hunk 1]
```diff
@@ -153,14 +157,17 @@ public function onDelete(Events\DynamicContentEvent $event)
...
     public function onTokenReplacement(MauticEvents\TokenReplacementEvent $event)
     {
-        /** @var Lead $lead */
         $lead         = $event->getLead();
         $content      = $event->getContent();
         $clickthrough = $event->getClickthrough();
 
-        if ($content) {
+        if ($lead instanceof Lead && $content) {
+            /** @var Lead $lead */
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php]: It is obvious that `$lead` variable must be an instance of `Lead` from the line above, right? So this comment is just stating obvious
```suggestion
```

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php [hunk 2]
```diff
@@ -90,7 +92,8 @@ public function __construct(
         DynamicContentHelper $dynamicContentHelper,
         DynamicContentModel $dynamicContentModel,
         CorePermissions $security,
-        ContactTracker $contactTracker
+        ContactTracker $contactTracker,
+        CompanyModel $companyModel
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php:96]:
```suggestion
        private CompanyModel $companyModel
```

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php [hunk 3]
```diff
@@ -79,6 +80,7 @@ class DynamicContentSubscriber implements EventSubscriberInterface
      * @var ContactTracker
      */
     private $contactTracker;
+    private CompanyModel $companyModel;
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php:83]:
```suggestion
```

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php [hunk 4]
```diff
@@ -102,6 +105,7 @@ public function __construct(
         $this->dynamicContentModel  = $dynamicContentModel;
         $this->security             = $security;
         $this->contactTracker       = $contactTracker;
+        $this->companyModel         = $companyModel;
```
Comment (escopecz) [inline app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php:108]:
```suggestion
```

File: app/bundles/LeadBundle/Entity/CompanyRepository.php [hunk 5]
```diff
@@ -126,12 +127,14 @@ public function getFieldGroups()
...
      * @return array
+     *
+     * @throws Exception
      */
     public function getCompaniesByLeadId($leadId, $companyId = null)
     {
         $q = $this->getEntityManager()->getConnection()->createQueryBuilder();
 
-        $q->select('comp.id, comp.companyname, comp.companycity, comp.companycountry, cl.is_primary')
+        $q->select('*')
```
Comment (escopecz) [inline app/bundles/LeadBundle/Entity/CompanyRepository.php]: This doesn't seem optimal. I guess you need all columns from the companies table. Do you need also all columns from the companies_leads table?

**Review by escopecz — CHANGES_REQUESTED** [review]

File: app/bundles/LeadBundle/Entity/CompanyRepository.php (hunk 5 above)
Comment (shinde-rahul) [reply app/bundles/LeadBundle/Entity/CompanyRepository.php]: Yes, I needed fields from the companies table and an `is_primary` field from companies_leads.

File: app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php (hunk 1 above)
Comment (shinde-rahul) [reply app/bundles/DynamicContentBundle/EventListener/DynamicContentSubscriber.php]: I am adding it to the patch. As of now, I don't recall why I have added that.

File: app/bundles/LeadBundle/Entity/CompanyRepository.php (hunk 5 above)
Comment (escopecz) [reply app/bundles/LeadBundle/Entity/CompanyRepository.php]: Would this work?
```suggestion
        $q->select('comp.*, cl.is_primary')
```

**Review by escopecz — APPROVED** [review]
The code looks good. I'd suggest to write a functional test instead of the unit test next time as a unit test doesn't guarantee that the feature works due to all the mocks that are just assumptions.

**Review by MadlenF — APPROVED** [review]
Works as described from user perspective!

**Comment by escopecz** [comment]
I have bad news. Since the M5 and M4.4.11 was released we do not accept any bug fixes to the M4. Only security fixes will be released to M4.

I confirmed this in https://mautic.slack.com/archives/CQMKV0RU1/p1706789913619169

This is the release plan: https://www.mautic.org/mautic-releases

**Pull request closed without merging**